presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,442 @@
+            nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+            offset += BITS_IN_BYTE;
+        }
+
+        // Process the remaining positions
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = block.isNull(positions[offset++]);
+        }
+    }
+
+    private void encodeNonNullsAsBits(int count)
+    {
+        if (remainingNullsCount + count < BITS_IN_BYTE) {
+            // just put all of this batch to remainingNullsFromLastBatch
+            for (int i = 0; i < count; i++) {
+                remainingNullsFromLastBatch[remainingNullsCount++] = false;
+            }
+            return;
+        }
+
+        int remainingPositions = count - encodeRemainingNullsAsBits();
+        int bytesCount = remainingPositions >>> 3;
+
+        nullsBufferIndex = fill(nullsBuffer, nullsBufferIndex, bytesCount, (byte) 0);
+
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = false;
+        }
+    }
+
+    private int encodeRemainingNullsAsBits()
+    {
+        if (remainingNullsCount == 0) {
+            return 0;
+        }
+
+        byte value = 0;
+        for (int i = 0; i < remainingNullsCount; i++) {
+            value |= remainingNullsFromLastBatch[i] ? 0b1000_0000 >>> i : 0;
+        }
+
+        nullsBufferContainsNull |= (value != 0);
+        nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+
+        int padding = BITS_IN_BYTE - remainingNullsCount;
+        remainingNullsCount = 0;
+        return padding;
+    }
+
+    protected long getNullsBufferSizeInBytes()
+    {
+        return nullsBufferIndex +                               // nulls buffer
+                remainingNullsCount;   // the remaining nulls not serialized yet
+    }
+
+    protected long getNullsBufferRetainedSizeInBytes()
+    {
+        return sizeOf(nullsBuffer) +                    // nulls buffer
+                sizeOf(remainingNullsFromLastBatch);    // the remaining nulls not serialized yet
+    }
+
+    protected long getNullsBufferSerializedSizeInBytes()
+    {
+        long size = SIZE_OF_BYTE;       // nulls uses 1 byte for mayHaveNull
+        if (containsNull()) {
+            size += nullsBufferIndex +  // nulls buffer
+                    (remainingNullsCount > 0 ? SIZE_OF_BYTE : 0);   // The remaining nulls not serialized yet. We have to add it here because at the time of calling this function, the remaining nulls was not put into the nullsBuffer yet.
+        }
+        return size;
+    }
+
+    protected long getPositionsSizeInBytes()
+    {
+        if (mappedPositions != null) {
+            return 2 * SIZE_OF_INT * positionCount;
+        }
+        return SIZE_OF_INT * positionCount;
+    }
+
+    protected long getPostionsRetainedSizeInBytes()
```

Comment:
typo: `i` is missing: getPostionsRetainedSizeInBytes  -> getPositionsRetainedSizeInBytes